File: src/main/java/com/tambapps/http/restclient/response/SuccessResponse.java
package com.tambapps.http.restclient.response;

import com.tambapps.http.restclient.response.handler.ResponseHandler;

public class SuccessResponse<T> extends AbstractRestResponse<T> {

  private final T data;

  public SuccessResponse(int responseCode, HttpHeaders headers, T data) {
    super(responseCode, headers);
    this.data = data;
  }

  @Override
  public boolean isErrorResponse() {
    return false;
  }

  @Override
  public boolean isSuccessful() {
    return true;
  }

  @Override
  public T getData() {
    return data;
  }

  @Override
  public <ErrorT> ErrorT getErrorData(ResponseHandler<ErrorT> responseHandler) {
    return null;
  }

  @Override
  public byte[] getRawErrorData() {
    return null;
  }
}

File: httpclient-core/src/main/kotlin/com/tambapps/http/client/response/SuccessResponse.kt
package com.tambapps.http.client.response

import com.tambapps.http.client.response.handler.ResponseHandler

class SuccessResponse<T> internal constructor(responseCode: Int, headers: HttpHeaders, private val myData: T?) : AbstractResponse<T>(responseCode, headers) {

    override val isErrorResponse = false
    override val isSuccessful = true
    override val hasData = myData != null
    override val data: T
        get() =  if (hasData)  myData!!
        else throw IllegalAccessException("Success response doesn't have error data")

    override fun <ErrorT> getErrorData(responseHandler: ResponseHandler<ErrorT>): ErrorT {
        throw IllegalAccessException("Success response doesn't have error data")
    }

    override val rawErrorData: ByteArray
        get() = throw IllegalAccessException("Success response doesn't have error data")

}